src/cascadia/PublicTerminalCore/HwndTerminalAutomationPeer.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "HwndTerminalAutomationPeer.hpp"
+#include "../../types/UiaTracing.h"
+#include <UIAutomationCoreApi.h>
+
+#pragma warning(suppress : 4471) // We don't control UIAutomationClient
+#include <UIAutomationClient.h>
+
+using namespace Microsoft::Console::Types;
+
+static constexpr wchar_t UNICODE_NEWLINE{ L'\n' };
+
+// Method Description:
+// - creates a copy of the provided text with all of the control characters removed
+// Arguments:
+// - text: the string we're sanitizing
+// Return Value:
+// - a copy of "sanitized" with all of the control characters removed
+static std::wstring Sanitize(std::wstring_view text)
+{
+    std::wstring sanitized{ text };
+    sanitized.erase(std::remove_if(sanitized.begin(), sanitized.end(), [](wchar_t c) {
+                        return (c < UNICODE_SPACE && c != UNICODE_NEWLINE) || c == 0x7F /*DEL*/;
+                    }),
+                    sanitized.end());
+    return sanitized;
+}
+
+// Method Description:
+// - verifies if a given string has text that would be read by a screen reader.
+// - a string of control characters, for example, would not be read.
+// Arguments:
+// - text: the string we're validating
+// Return Value:
+// - true, if the text is readable. false, otherwise.
+static constexpr bool IsReadable(std::wstring_view text)
+{
+    for (const auto c : text)
+    {
+        if (c > UNICODE_SPACE)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+void HwndTerminalAutomationPeer::RecordKeyEvent(const WORD vkey)
+{
+    if (const auto charCode{ MapVirtualKey(vkey, MAPVK_VK_TO_CHAR) })
+    {
+        if (const auto keyEventChar{ gsl::narrow_cast<wchar_t>(charCode) }; IsReadable({ &keyEventChar, 1 }))
+        {
+            _keyEvents.emplace_back(keyEventChar);
+        }
+    }
+}

```

Comment:
This is duplicated in the TermControlAutomationPeer. I think I'll move it over to the IUiaEventDispatcher so that it doesn't have to be duplicated

Reply:
But that's an interface. I don't think we should do that. Duplicate the code for now, fix later.